Mathlib/GroupTheory/GroupAction/Blocks.lean

```diff
@@ -0,0 +1,457 @@
+/-
+Copyright (c) 2024 Antoine Chambert-Loir. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Antoine Chambert-Loir
+
+-/
+
+import Mathlib.Data.Setoid.Partition
+import Mathlib.GroupTheory.GroupAction.Basic
+import Mathlib.GroupTheory.GroupAction.Pointwise
+import Mathlib.GroupTheory.GroupAction.SubMulAction
+import Mathlib.GroupTheory.Subgroup.Actions
+
+/-! # Blocks
+
+Given `SMul G X`, an action of a type `G` on a type `X`, we define
+
+- the predicate `IsBlock G B` states that `B : Set X` is a block,
+which means that the sets `g • B`, for `g ∈ G` form a partition of `X`.
+
+- a bunch of lemmas that give examples of “trivial” blocks : ⊥, ⊤, singletons,
+and non trivial blocks: orbit of the group, orbit of a normal subgroup…
+
+The non-existence of nontrivial blocks is the definition of primitive actions.
+
+## References
+
+We follow [wieland1964].
+
+-/
+
+open scoped BigOperators Pointwise
+
+namespace MulAction
+
+section orbits
+
+variable {G : Type*} [Group G] {X : Type*} [MulAction G X]
+
+theorem orbit.equal_or_disjoint (a b : X) :
```

Comment:
```suggestion
theorem orbit.eq_or_disjoint (a b : X) :
```
You could consider rephrasing this as `Pairwise Disjoint on (orbit G)`.

Reply:
Unless I'm mistaken with what Pairwise does, I am not sure that works: there are distinct points having the same orbit.

Reply:
Probably I am mistaken with what `Pairwise .. on` does.

Reply:
We want to say that the orbits are pairwise disjoint, but `Pairwise (Disjoint on (orbit G : X → Set X))` requires to prove that for `x y : X` such that `x ≠ y`, the sets `orbit G x` and `orbit G y` are disjoint.

Reply:
Do you prefer this:
```
example : (Set.range fun x : X => orbit G x).PairwiseDisjoint id := by
  rintro ox ⟨x, rfl⟩ oy ⟨y, rfl⟩ 
  unfold Disjoint
  rw [not_imp_comm]
  simp
  intro s hsy hsx
  erw [Classical.not_forall_not]
  rintro ⟨z, hz⟩
  rw [← orbit_eq_iff.mpr (hsx hz), ← orbit_eq_iff.mpr (hsy hz)]
```

Reply:
Or even that:
```
example : (Set.range fun x : X => orbit G x).PairwiseDisjoint id := by
  by_contra h 
  rcases Set.exists_ne_mem_inter_of_not_pairwiseDisjoint h with ⟨s, ⟨x, rfl⟩, t, ⟨y, rfl⟩, hst, z, ⟨hzx, hzy⟩⟩
  apply hst
  simp only [id_eq] at hzx hzy 
  simp only [← orbit_eq_iff.mpr hzx, ← orbit_eq_iff.mpr hzy]
```

Reply:
I added that one, and another proof for the fact that orbits of normal subgroups are blocks.
Do you feel they are more idomatic?